OSDOCS-6534: adding CSI snapshotting (#62557)
openshift/openshift-docs: modules/microshift-storage-creating-vol-snapshot.adoc

// Module included in the following assemblies:
//
// microshift/volume-snapshots-microshift.adoc

:_content-type: PROCEDURE
[id="creating-a-volume-snapshotting_{context}"]
= Creating a volume snapshot

To create a snapshot of a {product-title} storage volume, you must first configure {op-system-ostree} and the cluster. In the following example procedure, the pod that the source volume is mounted to is deleted. Deleting the pod prevents data continuing to be written to it during snapshot creation. If the pod is managed with a replication controller, meaning that a new pod is instantly created when one is deleted, temporarily scale the replica count to zero instead of deleting the pod.

[NOTE]
====
Volume snapshots are located on the same devices as the original data. To use the volume snapshots as backups, move the snapshots to a secure location.
====

.Prerequisites

> . User has root access to a {product-title} cluster.

Reviewer: Prereqs need to be unnumbered. I won't point this out any more if you can check the rest of the procedures.

. A {product-title} cluster is running.
. A device class defines an LVM thin-pool.
. A `volumeSnapshotClass` specifies `driver: topolvm.io`.
. Any workload attached to the source PVC is paused or deleted. This helps avoid data corruption.

[IMPORTANT]
====
All writes to the volume must be halted while you are creating the snapshot. If you do not halt writes, your data might be corrupted.
====

.Procedure

. Delete the pod to ensure that no data is written to the volume during snapshotting by running the following command:
+
[source, terminal]
----
$ oc delete my-pod
----

> . Optional: If the pod is managed with a replication controller, meaning that a new pod is instantly created when one is deleted, temporarily scale the replica count to zero instead of deleting the pod.

Reviewer: Someone might not see this option to scale the pod to zero until after they've already deleted the pod, since it's in a second step.

You could instead display it as two options, and the user chooses one.

```
. Prevent data from continuing to be written to the pod. (or maybe "Halt all writes to the volume" based on Michael's suggestion below_

* If the pod is managed with a replication controller, temporarily scale the ..."
* If the pod is not managed with a replication controller, delete the pod ...
```

> . Optional: If the pod is managed with a replication controller, meaning that a new pod is instantly created when one is deleted, temporarily scale the replica count to zero instead of deleting the pod.

Reviewer 2: This could just be a note as part of step 1, since it's an alternative to that step if they are using their own workload instead of our example.

. Once the pod has been deleted, run a command similar to the example that follows. Insert your own configuration details.
+
.Example snapshot configuration
+
[source, terminal]
----

> # oc apply -f <<EOF

Reviewer 2: We should scan the shell prompts in these procedures to make sure they're consistent. This one is for root, the next one is for a regular user. I don't think the user needs to be root for any of these steps, right @copejon ?

Reviewer 3: Yes. And agreed - we should stick to non-root command examples.

Author: I made these all `$`. Let me know if we need to stick a sudo in them anywhere.

apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshot # <1>
metadata:

>   name: <my-snap> # <2>

Reviewer: If it was an example, then "my-snap" would be fine, without the < >. But if it's a replaceable value, then I'd recommend something like "<snapshot_name>"

The rest of your example uses example values without < >, so probably could just remove the < > here

spec:
  volumeSnapshotClassName: topolvm-snapclass # <3>
  source:
    persistentVolumeClaimName: test-claim-thin # <4>
EOF
----
[.small]

> <1> Create a `VolumeSnapshot` object.

Reviewer: This callout is probably unnecessary since it's pretty obvious

<2> The name that you specify for the snapshot.
<3> Specify the desired `VolumeSnapshotClass`.

> <4> Specify either `persistentVolumeClaimName` or `volumeSnapshotContentName`. In this example, we are creating a snapshot from a PVC. The value `test-claim-thin` is the source PVC.

Reviewer: Avoid "we"

```suggestion
<4> Specify either `persistentVolumeClaimName` or `volumeSnapshotContentName`. In this example, the snapshot is created from a PVC named `test-claim-thin`.
```

> <4> Specify either `.persistentVolumeClaimName` or `volumeSnapshotContentName`. In this example, we are creating a snapshot from a PVC. The value `test-claim-thin` is the source PVC.

Reviewer 2: ```suggestion
<4> Specify either `persistentVolumeClaimName` or `volumeSnapshotContentName`. In this example, we are creating a snapshot from a PVC. The value `test-claim-thin` is the source PVC.
```

I think the `.` was just a typo, but it makes it look like we're giving inconsistent syntax.

. Wait for the storage driver to finish creating the snapshot by running the following command:
+
[source, terminal]
----
$ oc wait volumesnapshot/my-snap --for=jsonpath\='{.status.readyToUse}=true'
----

. When the `volumeSnapshot` is `ReadyToUse`, you can restore it as a volume for future PVCs.

> . After you have created the volume snapshot, you can remount the source PVC to a new pod.

Reviewer: Is this the step where they would re-create the pod (if they deleted it), or scale the replica count back up (if they scaled it to zero)? If so, I'd be more explicit about that.

> . To use the volume snapshot as a backup, move the snapshot to a secure location.

Reviewer 2: @copejon, what form does the snapshot take and how do they access it to copy it off of the host?

Reviewer 2: Yes, I think we want the steps for identifying the volume with the snapshot on it. Let's see if we can use the `-o` option to `oc` for extracting the field value instead of requiring `yq`.

Once the user has the volume identity, the way they access the content is just to mount it and copy the files out, right @copejon ?

Reviewer 3: > oc -c 
+1 - my rushed example shouldn't be taken as cannon :D 

> Once the user has the volume identity, the way they access the content is just to mount it and copy the files out

yep, exactly. That's something we'll need to add to the downstream repo sooner rather than later

Author: I made a new module that I will put in this PR, but it surely needs engineering insight, so it's also in the google doc. I'll mark this string resolved and we can take it up as a new topic in either place.
